### PR TITLE
Ensure findPrecedingToken recurses into JSDoc children when needed

### LINKED ISSUE #53476: Infinite loop getting completions when finding preceding matching token in JSDoc
# Bug Report

The infinite loop happens here:

https://github.com/microsoft/TypeScript/blob/d5fd34b03718b851adad6c25f7de880ef20173b3/src/services/utilities.ts#L1935-L1952

More specifically, the following function returns back the current token:

https://github.com/microsoft/TypeScript/blob/9bd1a3225be540e90e230d4cfe9577673ac1e2c9/src/services/utilities.ts#L1700

### 🔎 Search Terms

infinite loop completions jsdoc

### 🕗 Version & Regression Information

5.0.2 and I can reproduce it in main. It probably occurred in older versions.

### ⏯ Playground Link

I can't reproduce it in the playground.

### 💻 Code

1. Take the following starting code:

```ts
/// <reference path="./compiler.d.ts" />

(window => {
  /** @type {DenoCore} */
  const core = window.Deno.core;
})();
```

2. Type `/** @` before `window`:

```ts
/// <reference path="./compiler.d.ts" />

(/** @window => {
  /** @type {DenoCore} */
  const core = window.Deno.core;
})();
```

3. Hover over `window`.

### 🙁 Actual behavior

Hangs forever.

### 🙂 Expected behavior

Not hang.

## Test case

```ts
import * as ts from "../../_namespaces/ts";

describe("unittests:: services:: utilities", () => {
  describe("Test findPrecedingMatchingToken,", () => {
    it("should not infinite loop finding opening brace", () => {
      const sourceFile = ts.createSourceFile("file.ts", `/// <reference path="./compiler.d.ts" />

(/** @window => {
  /** @type {Abcd123} */
  const core = window.Abcd.core;
})();`, ts.ScriptTarget.ESNext, true);
      // can't use ts.getTokenAtPosition because it returns back the wrong token
      const param = ts.forEachChildRecursively(sourceFile, node => node.kind === ts.SyntaxKind.Parameter ? node : undefined)!;
      const jsDoc = param.getChildren()[0];
      const token = jsDoc.getLastToken()!;
      const result = ts.findPrecedingMatchingToken(token, ts.SyntaxKind.OpenBraceToken, sourceFile);
      // will never happen
      console.log(result);
    });
  });
});
```

I was looking into this a bit, but I need to move on to other things so I'm just logging it.

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #53476

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot perf test this
@typescript-bot pack this

File: src/services/utilities.ts [hunk 1]
```diff
@@ -1739,7 +1739,14 @@ export function findPrecedingToken(position: number, sourceFile: SourceFileLike,
                 if (lookInPreviousChild) {
                     // actual start of the node is past the position - previous token should be at the end of previous child
                     const candidate = findRightmostChildNodeWithTokens(children, /*exclusiveStartPosition*/ i, sourceFile, n.kind);
-                    return candidate && findRightmostToken(candidate, sourceFile);
+                    if (candidate) {
+                        // Ensure we recurse into JSDoc nodes with children.
+                        if (isJSDocCommentContainingNode(candidate) && candidate.getChildren(sourceFile).length) {
```
Comment (jakebailey) [inline src/services/utilities.ts:1744]: Checking the children length mirrors the change in #39912 and is needed to keep its test passing.

I'd use `getChildCount` but surprise, that only accepts a `SourceFile`, and in this helper we only have `SourceFileLike`. Seems like an oops.

**Comment by DanielRosenwasser** [comment]
What caused the regression in the first place?

**Comment by jakebailey** [comment]
#53476 says

> 5.0.2 and I can reproduce it in main. It probably occurred in older versions.

I can test 4.9 to double check, though.

**Comment by jakebailey** [comment]
It hangs in 4.9 too, so this isn't a regression in 5.0 as far as I can tell.

File: src/testRunner/unittests/services/utilities.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    describe("Test findPrecedingMatchingToken,", () => {
+        it("should not infinite loop finding opening brace", () => {
+            const sourceFile = ts.createSourceFile("file.ts", `/// <reference path="./compiler.d.ts" />
+
+(/** @window => {
+  /** @type {Abcd123} */
+  const core = window.Abcd.core;
+})();`, ts.ScriptTarget.ESNext, /*setParentNodes*/ true);
+            // can't use ts.getTokenAtPosition because it returns back the wrong token
+            const param = ts.forEachChildRecursively(sourceFile, node => node.kind === ts.SyntaxKind.Parameter ? node : undefined)!;
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/services/utilities.ts:13]: what parameter is there in the example code?

File: src/testRunner/unittests/services/utilities.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import * as ts from "../../_namespaces/ts";
+
+describe("unittests:: services:: utilities", () => {
+    describe("Test findPrecedingMatchingToken,", () => {
+        it("should not infinite loop finding opening brace", () => {
+            const sourceFile = ts.createSourceFile("file.ts", `/// <reference path="./compiler.d.ts" />
+
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/services/utilities.ts:7]: do you need the triple slash reference?

File: src/testRunner/unittests/services/utilities.ts (hunk 2 above)
Comment (DanielRosenwasser) [reply src/testRunner/unittests/services/utilities.ts:13]: oh, `window`??

File: src/testRunner/unittests/services/utilities.ts (hunk 3 above)
Comment (jakebailey) [reply src/testRunner/unittests/services/utilities.ts:7]: No, probably not.

File: src/testRunner/unittests/services/utilities.ts (hunk 2 above)
Comment (jakebailey) [reply src/testRunner/unittests/services/utilities.ts:13]: No idea, I just copied and pasted the test from the issue...

**Review by sandersn — APPROVED** [review]

**Pull request merged**